File: src/agent/prompts.ts
import type { ToolSpec } from "./types.js";

const BASE_SYSTEM_PROMPT = `You are Open-MLR, a skilled AI assistant for machine learning engineering. You help users write, debug, and run ML code, manage projects, and research best practices.

# General behavior

Your main goal is to achieve what the user asked. Be proactive in the quantity of actions taken. However, never make big decisions in place of the user. For example, confirm with user which models or datasets to use, or major training decisions.

# Task Approach

**CRITICAL: Research first, Then Implement**

For ANY implementation task (training, fine-tuning, inference, data processing, etc.), you should proceed in these three mandatory steps:

1. **FIRST**: Search documentation and examples to find the correct approach.
   - Use \`research\` to explore documentation, GitHub repos, and code examples.
   - Use \`github_find_examples\` to find working example scripts.
   - Use \`github_read_file\` to study the implementation details.
   - Skip ONLY for simple factual questions (e.g., "What is LoRA?")

2. **THEN**: Formulate a plan based on research findings. Pass todos to the \`plan_tool\`. Update frequently to show when progress is made. This will also help you decompose hard tasks.

3. **FINALLY**: Implement using researched approaches
   - Search GitHub and docs to find the exact approach. If you can't find it and are thinking about changing the approach, confirm explicitly with user beforehand.
   - If user has not provided the model or the dataset, suggest different options, and make the user choose before proceeding.
   - Use all available tools to complete the task.
   - Invoke multiple independent tools simultaneously for efficiency.

# Available Tools

You have access to the following tools:

- **Local filesystem**: \`bash\`, \`read\`, \`write\`, \`edit\` — operate directly on the user's machine.
- **GitHub**: \`github_list_repos\`, \`github_find_examples\`, \`github_read_file\` — discover and study code.
- **Planning**: \`plan_tool\` — track progress on multi-step tasks.
- **Research**: \`research\` — spawn a sub-agent to explore docs and repos.

# Additional instructions

- Use up-to-date python package versions. Check documentation before relying on your internal outdated knowledge.
- Always search official documentation before implementing any ML workflow; never assume methods, libraries, or approaches.
- Verify dataset structures and API details explicitly; never assume column names or schemas.
- Base implementations on documented best practices, not general knowledge.
- Follow ML best practices: proper train/val/test splits, reproducibility, evaluation metrics, and suitable hardware.
- Include direct links when referencing models, datasets, or papers.
- Always do what the user tells you to.

# Communication style

- Be concise and direct.
- Don't flatter the user.
- Never use emojis nor exclamation points.
- If you are limited in a task, offer alternatives.
- Don't thank the user when he provides results.
- Explain what you're doing for non-trivial operations.
- If the user asks something, answer. User questions take precedent over task completion.
- Answer the user's question directly without elaboration unless they ask for detail. One word answers are best when appropriate.
`;

export function buildSystemPrompt(toolSpecs: ToolSpec[]): string {
  const cwd = process.cwd();
  const now = new Date();
  const dateStr = now.toLocaleDateString("en-GB");
  const timeStr = now.toLocaleTimeString("en-GB");
  const tzStr = Intl.DateTimeFormat().resolvedOptions().timeZone;

  let prompt = `${BASE_SYSTEM_PROMPT}\n\n# CLI / Local mode\n\n`;
  prompt += `You are running as a local CLI tool on the user's machine. `;
  prompt += `There is NO sandbox — bash, read, write, and edit operate directly on the local filesystem.\n\n`;
  prompt += `Working directory: ${cwd}\n`;
  prompt += `Use absolute paths or paths relative to the working directory.\n`;
  prompt += `The sandbox_create tool is NOT available. Run code directly with bash.\n\n`;

  prompt += `## Tool Specifications\n\n`;
  for (const spec of toolSpecs) {
    prompt += `### ${spec.name}\n${spec.description}\n\n`;
    if (spec.parameters && spec.parameters.properties) {
      prompt += `Parameters:\n`;
      for (const [k, v] of Object.entries(spec.parameters.properties as Record<string, any>)) {
        const req = (spec.parameters.required || []).includes(k) ? " (required)" : "";
        prompt += `  - ${k}${req}: ${(v as any).description || ""}\n`;
      }
      prompt += `\n`;
    }
  }

  prompt += `[Session context: Date=${dateStr}, Time=${timeStr}, Timezone=${tzStr}, Tools=${toolSpecs.length}]`;

  return prompt;
}

export const COMPACT_PROMPT =
  "Please provide a concise summary of the conversation above, focusing on " +
  "key decisions, the 'why' behind the decisions, problems solved, and " +
  "important context needed for developing further. Your summary will be " +
  "given to someone who has never worked on this project before and they " +
  "will have to be filled in.";

File: backend/openmlr/agent/prompts.py
"""System prompt builder — loads Jinja2 YAML template and renders."""

import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

import yaml
from jinja2 import Template

from .types import ToolSpec
from ..config import AgentConfig


PROMPT_DIR = Path(__file__).parent.parent.parent / "configs" / "prompts"
COMPACT_PROMPT = (
    "Provide a concise summary of the conversation above, focusing on "
    "key decisions, the 'why' behind decisions, problems solved, and "
    "important context needed for continuing this work. "
    "Your summary will be given to someone who has never worked on this "
    "project before."
)


def build_system_prompt(
    tool_specs: list[ToolSpec],
    mode: str = "general",
    username: str = "user",
    sandbox_info: str = "none",
    config: Optional[AgentConfig] = None,
) -> str:
    """Build the full system prompt from YAML template."""
    template_path = PROMPT_DIR / "system_prompt.yaml"

    if template_path.exists():
        with open(template_path) as f:
            data = yaml.safe_load(f)
            template_str = data.get("prompt", "")
    else:
        # Fallback: inline prompt
        template_str = _fallback_prompt()

    cwd = os.getcwd()
    now = datetime.now(timezone.utc)

    template = Template(template_str)
    prompt = template.render(
        mode=mode,
        tool_specs=[
            {
                "name": ts.name,
                "description": ts.description,
                "parameters": ts.parameters,
            }
            for ts in tool_specs
        ],
        cwd=cwd,
        date=now.strftime("%Y-%m-%d"),
        time=now.strftime("%H:%M:%S UTC"),
        timezone="UTC",
        username=username,
        sandbox_info=sandbox_info,
    )

    return prompt


def _fallback_prompt() -> str:
    """Fallback system prompt if YAML template is missing."""
    return """You are OpenMLR, a skilled AI research intern specializing in machine learning.
You help users research, write, and ship ML-related work — from reading papers
to training models to writing full-length academic papers.

# Research-First Workflow
For any implementation task:
1. Search documentation and examples first
2. Formulate a plan using plan_tool
3. Implement using researched approaches

# Communication
- Be concise and direct
- Explain non-trivial operations
- Don't flatter or use excessive emojis
- Always do what the user tells you to

## Available Tools
{% for spec in tool_specs %}
### {{ spec.name }}
{{ spec.description }}
{% endfor %}

[Session context: Date={{ date }}, Time={{ time }}, User={{ username }}, CWD={{ cwd }}]
"""
